File: examples/components/services/src/branch.rs
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

//! This program launches `fuchsia.examples.services.BankAccount` service providers and consumes
//! their instances.
//!
//! This program is written as a test so that it can be easily launched with `fx test`.

use {
    fidl_fuchsia_examples_services as fexamples, fidl_fuchsia_io as fio,
    fidl_fuchsia_sys2 as fsys2, fuchsia_component::client as fclient, log::*,
};

const COLLECTION_NAME: &'static str = "account_providers";
const TEST_PACKAGE: &'static str = "fuchsia-pkg://fuchsia.com/service-examples";

#[fuchsia::test]
async fn read_and_write_to_multiple_service_instances() {
    // Launch two BankAccount providers into the `account_providers` collection.
    let realm = fuchsia_component::client::connect_to_protocol::<fsys2::RealmMarker>()
        .expect("connect to Realm service");
    start_provider(&realm, "a", &format!("{}#meta/provider-a.cm", TEST_PACKAGE)).await;
    start_provider(&realm, "b", &format!("{}#meta/provider-b.cm", TEST_PACKAGE)).await;

    let service_dir = fclient::open_service::<fexamples::BankAccountMarker>()
        .expect("failed to open service dir");
    let instances = files_async::readdir(&service_dir)
        .await
        .expect("failed to read entries from service_dir")
        .into_iter()
        .map(|dirent| dirent.name);

    // Debit both bank accounts by $5.
    for instance in instances {
        let proxy = fclient::connect_to_service_instance::<fexamples::BankAccountMarker>(&instance)
            .expect("failed to connect to service instance");
        let read_only_account = proxy.read_only().expect("read_only protocol");
        let owner = read_only_account.get_owner().await.expect("failed to get owner");
        let initial_balance = read_only_account.get_balance().await.expect("failed to get_balance");
        info!("retrieved account for owner '{}' with balance ${}", &owner, &initial_balance);

        let read_write_account = proxy.read_write().expect("read_write protocol");
        assert_eq!(read_write_account.get_owner().await.expect("failed to get_owner"), owner);
        assert_eq!(
            read_write_account.get_balance().await.expect("failed to get_balance"),
            initial_balance
        );
        info!("debiting account of owner '{}'", &owner);
        read_write_account.debit(5).await.expect("failed to debit");
        assert_eq!(
            read_write_account.get_balance().await.expect("failed to get_balance"),
            initial_balance - 5
        );
    }
}

async fn start_provider(realm: &fsys2::RealmProxy, name: &str, url: &str) -> fio::DirectoryProxy {
    info!("creating BankAccount provider \"{}\" with url={}", name, url);
    let child_args =
        fsys2::CreateChildArgs { numbered_handles: None, ..fsys2::CreateChildArgs::EMPTY };
    realm
        .create_child(
            &mut fsys2::CollectionRef { name: COLLECTION_NAME.to_string() },
            fsys2::ChildDecl {
                name: Some(name.to_string()),
                url: Some(url.to_string()),
                startup: Some(fsys2::StartupMode::Lazy),
                environment: None,
                ..fsys2::ChildDecl::EMPTY
            },
            child_args,
        )
        .await
        .expect("failed to make create_child FIDL call")
        .expect("failed to create_child");

    let (exposed_dir, exposed_dir_server_end) =
        fidl::endpoints::create_proxy::<fio::DirectoryMarker>()
            .expect("failed to create endpoints");
    info!("open exposed dir of BankAccount provider \"{}\" with url={}", name, url);
    realm
        .open_exposed_dir(
            &mut fsys2::ChildRef {
                name: name.to_string(),
                collection: Some(COLLECTION_NAME.to_string()),
            },
            exposed_dir_server_end,
        )
        .await
        .expect("failed to make open_exposed_dir FIDL call")
        .expect("failed to open_exposed_dir");
    exposed_dir
}

File: examples/components/services/cpp/branch.cc
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <fuchsia/component/cpp/fidl.h>
#include <fuchsia/component/decl/cpp/fidl.h>
#include <fuchsia/examples/services/cpp/fidl.h>
#include <lib/sys/cpp/component_context.h>
#include <lib/sys/service/cpp/service.h>
#include <lib/sys/service/cpp/service_aggregate.h>
#include <lib/syslog/cpp/log_settings.h>
#include <lib/syslog/cpp/macros.h>

#include <gtest/gtest.h>
#include <src/lib/testing/loop_fixture/real_loop_fixture.h>

class BankAccountTest : public ::gtest::RealLoopFixture {
 public:
  explicit BankAccountTest() {
    auto context = sys::ComponentContext::Create();
    // Connect to the fuchsia.component.Realm framework protocol
    context->svc()->Connect(realm_proxy_.NewRequest());
    // Return a channel connected to the /svc directory.
    svc_ = context->svc()->CloneChannel();
  }

  // Create and instance of the provider component and return its exposed directory
  fuchsia::io::DirectorySyncPtr StartProvider(const char* component_name,
                                              const char* component_url) {
    FX_SLOG(INFO, "Creating BankAccount provider", KV("url", component_url),
            KV("name", component_name));
    fuchsia::component::decl::CollectionRef collection_ref = {
        .name = "account_providers",
    };
    fuchsia::component::decl::Child child_decl;
    child_decl.set_name(component_name);
    child_decl.set_url(component_url);
    child_decl.set_startup(fuchsia::component::decl::StartupMode::LAZY);

    fuchsia::component::Realm_CreateChild_Result create_result;
    realm_proxy_->CreateChild(std::move(collection_ref), std::move(child_decl),
                              fuchsia::component::CreateChildArgs(), &create_result);

    FX_SLOG(INFO, "Open exposed dir of BankAccount provider", KV("url", component_url),
            KV("name", component_name));
    fuchsia::component::decl::ChildRef child_ref = {
        .name = component_name,
        .collection = "account_providers",
    };
    fuchsia::io::DirectorySyncPtr exposed_dir;
    fuchsia::component::Realm_OpenExposedDir_Result open_result;
    realm_proxy_->OpenExposedDir(child_ref, exposed_dir.NewRequest(), &open_result);

    return exposed_dir;
  }

 protected:
  // TODO(https://fxbug.dev/101092): Use default namespace to connect services.
  fidl::InterfaceHandle<fuchsia::io::Directory> svc_;

 private:
  fuchsia::component::RealmSyncPtr realm_proxy_;
};

TEST_F(BankAccountTest, ReadWriteMultipleServiceInstances) {
  // Launch two BankAccount providers into the `account_providers` collection.
  StartProvider("a", "#meta/provider-a.cm");
  StartProvider("b", "#meta/provider-b.cm");

  // List available instances of the BankAccount service
  auto service_aggregate =
      sys::OpenServiceAggregateAt<fuchsia::examples::services::BankAccount>(svc_);
  auto instance_names = service_aggregate.ListInstances();
  ZX_ASSERT(!instance_names.empty());

  // Debit both bank accounts by $5.
  for (auto& instance : instance_names) {
    auto service = sys::OpenServiceAt<fuchsia::examples::services::BankAccount>(svc_, instance);
    // Connect to the ReadOnlyAccount protocol offered by this service
    fuchsia::examples::services::ReadOnlyAccountSyncPtr read_only_account =
        service.read_only().Connect().BindSync();
    std::string initial_owner;
    read_only_account->GetOwner(&initial_owner);
    int64_t initial_balance;
    read_only_account->GetBalance(&initial_balance);
    FX_LOGS(INFO) << "Retrieved account: " << initial_owner;

    // Connect to the ReadWriteAccount protocol offered by this service
    fuchsia::examples::services::ReadWriteAccountSyncPtr read_write_account =
        service.read_write().Connect().BindSync();
    std::string owner;
    read_write_account->GetOwner(&owner);
    ASSERT_EQ(initial_owner, owner);
    int64_t balance;
    read_write_account->GetBalance(&balance);
    ASSERT_EQ(initial_balance, balance);

    FX_LOGS(INFO) << "Debiting account: " << owner;
    bool success;
    read_write_account->Debit(5, &success);
    ASSERT_TRUE(success);
    read_write_account->GetBalance(&balance);
    ASSERT_EQ(initial_balance - 5, balance);
  }
}
